src/test/app/Batch_test.cpp

```diff
@@ -0,0 +1,3636 @@
+        auto const bob = Account("bob");
+        env.fund(XRP(10000), alice, bob);
+        env.close();
+
+        STTx const stx = STTx(ttAMENDMENT, [&](auto& obj) {
+            obj.setAccountID(sfAccount, AccountID());
+            obj.setFieldH256(sfAmendment, uint256(2));
+            obj.setFieldU32(sfLedgerSequence, env.seq(alice));
+            obj.setFieldU32(sfFlags, tfInnerBatchTxn);
+        });
+
+        std::string reason;
+        BEAST_EXPECT(isPseudoTx(stx));
+        BEAST_EXPECT(!passesLocalChecks(stx, reason));
+        BEAST_EXPECT(reason == "Cannot submit pseudo transactions.");
+        env.app().openLedger().modify([&](OpenView& view, beast::Journal j) {
+            auto const result = ripple::apply(env.app(), view, stx, tapNONE, j);
+            BEAST_EXPECT(!result.applied && result.ter == temINVALID_FLAG);
+            return result.applied;
+        });
+    }
+
+    void
+    testOpenLedger(FeatureBitset features)
+    {
+        testcase("batch open ledger");
+        // IMPORTANT: When a transaction is submitted outside of a batch and
+        // another transaction is part of the batch, the batch might fail
+        // because the sequence is out of order. This is because the canonical
+        // order of transactions is determined by the account first. So in this
+        // case, alice's batch comes after bobs self submitted transaction even
+        // though the payment was submitted after the batch.
+
+        using namespace test::jtx;
+        using namespace std::literals;
+
+        test::jtx::Env env{*this, envconfig()};
+        XRPAmount const baseFee = env.current()->fees().base;
+
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+
+        env.fund(XRP(10000), alice, bob);
+        env.close();
+
+        env(noop(bob), ter(tesSUCCESS));
+        env.close();
+
+        auto const aliceSeq = env.seq(alice);
+        auto const preAlice = env.balance(alice);
+        auto const preBob = env.balance(bob);
+        auto const bobSeq = env.seq(bob);
+
+        // Alice Pays Bob (Open Ledger)
+        auto const payTxn1 = env.jt(pay(alice, bob, XRP(10)), seq(aliceSeq));
+        auto const payTxn1ID = to_string(payTxn1.stx->getTransactionID());
+        env(payTxn1, ter(tesSUCCESS));
+
+        // Alice & Bob Atomic Batch
+        auto const batchFee = batch::calcBatchFee(env, 1, 2);
+        auto const [txIDs, batchID] = submitBatch(
+            env,
+            tesSUCCESS,
+            batch::outer(alice, aliceSeq + 1, batchFee, tfAllOrNothing),
+            batch::inner(pay(alice, bob, XRP(10)), aliceSeq + 2),
+            batch::inner(pay(bob, alice, XRP(5)), bobSeq),
+            batch::sig(bob));
+
+        // Bob pays Alice (Open Ledger)
+        auto const payTxn2 = env.jt(pay(bob, alice, XRP(5)), seq(bobSeq + 1));
+        auto const payTxn2ID = to_string(payTxn2.stx->getTransactionID());
+        env(payTxn2, ter(terPRE_SEQ));
```

Comment:
is this txn ever successful? is it ever being retried in the next ledger? would be clarify the end result of this txn

Reply:
added: https://github.com/XRPLF/rippled/pull/5060/commits/ac54f33de3c05d205b803b2092e918ffc5094399